modules/pam_setquota/pam_setquota.c

```diff
@@ -0,0 +1,252 @@
+/* PAM setquota module
+
+   This PAM module sets disk quota when a session begins.
+
+   Copyright © 2006 Ruslan Savchenko <savrus@mexmat.net>
+   Copyright © 2010 Shane Tzen <shane@ict.usc.edu>
+   Copyright © 2012 Sven Hartge <sven@svenhartge.de>
+   Copyright © 2016 Keller Fuchs <kellerfuchs@hashbang.sh>
+
+*/
+
+#include <sys/types.h>
+#include <sys/quota.h>
+#include <linux/quota.h>
+#include <pwd.h>
+#include <syslog.h>
+#include <errno.h>
+#include <mntent.h>
+#include <stdio.h>
+
+#define PAM_SM_SESSION
+
+#include <security/pam_modules.h>
+#include <security/_pam_macros.h>
+#include <security/pam_ext.h>
+#include <security/pam_modutil.h>
+
+#define LOGIN_DEFS "/etc/login.defs"
+
+#ifndef TRUE
+# define TRUE 1
+#endif
+#ifndef FALSE
+# define FALSE 0
+#endif
+
+struct pam_params {
+  uid_t start_uid;
+  uid_t end_uid;
+  char *fs;
+  int overwrite;
+  int debug;
+};
+
+static inline void debug(pam_handle_t *pamh, struct if_dqblk *p) {
+  pam_syslog(pamh, LOG_DEBUG, "bsoftlimit=%llu bhardlimit=%llu"
+                              "isoftlimit=%llu ihardlimit=%llu",
+             p->dqb_bsoftlimit, p->dqb_bhardlimit, p->dqb_isoftlimit,
+             p->dqb_ihardlimit);
+}
+
+static int _pam_parse_dqblk(int argc, const char **argv, struct if_dqblk *p) {
+  int bhard = FALSE, bsoft = FALSE, ihard = FALSE, isoft = FALSE;
+
+  for (; argc-- > 0; ++argv) {
+    if (strncmp(*argv, "bhardlimit=", 11) == 0) {
+      p->dqb_bhardlimit = strtol(*argv + 11, NULL, 10);
+      p->dqb_valid |= QIF_BLIMITS;
+      bhard = TRUE;
+    } else if (strncmp(*argv, "bsoftlimit=", 11) == 0) {
+      p->dqb_bsoftlimit = strtol(*argv + 11, NULL, 10);
+      p->dqb_valid |= QIF_BLIMITS;
+      bsoft = TRUE;
+    } else if (strncmp(*argv, "ihardlimit=", 11) == 0) {
+      p->dqb_ihardlimit = strtol(*argv + 11, NULL, 10);
+      p->dqb_valid |= QIF_ILIMITS;
+      ihard = TRUE;
+    } else if (strncmp(*argv, "isoftlimit=", 11) == 0) {
+      p->dqb_isoftlimit = strtol(*argv + 11, NULL, 10);
+      p->dqb_valid |= QIF_ILIMITS;
+      isoft = TRUE;
+    } else if (strncmp(*argv, "btime=", 6) == 0) {
+      p->dqb_btime = strtol(*argv + 6, NULL, 10);
+      p->dqb_valid |= QIF_BTIME;
+    } else if (strncmp(*argv, "itime=", 6) == 0) {
+      p->dqb_itime = strtol(*argv + 6, NULL, 10);
+      p->dqb_valid |= QIF_ITIME;
+    }
+  }
+
+  return !(bhard ^ bsoft) && !(ihard ^ isoft);
+}
+
+static void _pam_parse_params(int argc, const char **argv,
+                              struct pam_params *p) {
+  for (; argc-- > 0; ++argv) {
+    if (strncmp(*argv, "startuid=", 9) == 0)
+      p->start_uid = strtol(*argv + 9, NULL, 10);
+    else if (strncmp(*argv, "enduid=", 7) == 0)
+      p->end_uid = strtol(*argv + 7, NULL, 10);
+    else if (strncmp(*argv, "fs=", 3) == 0)
+      p->fs = (char *)*argv + 3;
```

Comment:
The cast from (const char *) to (char *) looks really bad.

Reply:
That code is going to be replaced with the new argv parsing using pam_str_skip_prefix anyways, isn't it?

Reply:
Yes, but the problem with the trailing slash will have to be dealt with anyway.

Reply:
Right. Which is quite possibly the reason the original author made it char instead of const char. To be able to manipulate the string later on.

Reply:
I have moved the /-removal function into _pam_parse_params, but because I have to change the fs parameter if it has a trailing /, I need to cast something somewhere. My solution so far is at https://github.com/linux-pam/linux-pam/pull/94/files#diff-a68500e7c99fcd33f0d387c65f0257f7R100 but I am open to anything more sophisticated.

Reply:
Nix that, casting to const char* is not necessary, only the other way around.